lib/pure/asynchttpserver.nim

```diff
@@ -34,9 +34,8 @@
 ##
 ##     server.listen Port(5555)
 ##     while true:
-##       if server.shouldAcceptRequest(5):
-##         var (address, client) = await server.socket.acceptAddr()
-##         asyncCheck processClient(server, client, address, cb)
+##       if server.shouldAcceptRequest():
+##         asyncCheck server.acceptRequest(cb)
 ##       else:
 ##         poll()
```

Comment:
Please revert this back to the default example and get rid of this default behaviour.

Reply:
No. Why? It's good code.

Reply:
Because until we establish a real life scenario where this matters it is just another thing that can hide bugs. If you have an FD leak this will hide it and you'll just have people wondering why their server is not responding to requests fast enough.

Reply:
For fd leak testing, people can still pass 0 (zero), no?  Maybe -1?  and/or even better do `setrlimit`.  A (slightly) slower but still working server might be a "performance bug". Such seem secondary to correctness bugs.  This also adds but 1 more out of dozens to hundreds of possible sources of performance mystery.  So, that impact is marginal, at best. Your alternative `safeAccept` idea could also similarly hide leaks.

Reply:
Also, the range of default open file limits is huge. MacOSX is like 256 while FreeBSD is 117036 and windows is 16M.  So, your OS/sysadmin can (and does) undermine your leak detection ability anyway -- unless, of course, you do `setrlimit` which is the recommended way to find leaks early/fail fast. If you are already leak-hunting/calling that, changing the arg to 0 seems not much extra work.  If you are not leak hunting, you would prefer things to just keep working under whatever load.

Why, taking your argument to its logical conclusion lands at the Nim runtime doing some `setrlimit(RLIMIT_FILE, 8)` on behalf of all programs even though it's almost impossible to know "8" is a good default for any given program.  The OSes/admins are just making wild guesses, too, as @Araq mentioned in some other thread earlier.  Any real leak detection scheme requires some prior knowledge of how many files should be open, and only the program knows that.  This is probably why Unix has evolved from tiny limits to giant limits.  Load/need variation was too hard to anticipate.

Reply:
You're blowing past the fact that I am still asking for a real life repro for this bug that isn't a synthetic case. You said yourself that the OS manages the ``accept`` queue anyway, so having this logic just reimplements this logic in a non-standard way.

Reply:
You seem to be not understanding that the listen backlog does not solve this.  I don't know what to say about that.  You need at least one fd free for accept to succeed.  Once you solve that problem, why not "over solve" it to achieve more robustness?

In terms of a real life example, diversity in connection lifetimes alone can cause pushing against fd limits.  Some connections last hours while others last milliseconds.  You get a big pool of idle or nearly idle/hung connections.  Real life example constructed.  Done.  This example is actually where the [C10k](https://en.wikipedia.org/wiki/C10k_problem) problem came from with many dial-up modems/elsewise high latency clients to ftp servers.  ftp has a stay open and sit idle shell.  That problem, in turn, was what inspired epoll/kqueue style interfaces/various OS scalability.  That in turn is why there is a lot of example code out there with `wrk -c 1000` which is probably where @mrhdias got his from.

You insist on a "real" repro when we had a "real" forum person trip over this and beg for a quick solution and be thankful he got one.  There are simple arguments about how the problem can arise.  There is massive 64 to 16M range in limits.  Your insistence is seeming enough like willful stubbornness, TBH, that it demotivates me from further trying to get you to understand.  Have a nice weekend!

Reply:
> You insist on a "real" repro when we had a "real" forum person trip over this and beg for a quick solution and be thankful he got one. 

I just want to establish exactly what the forum person tripped over. It's simply not clear from their post, everybody here is inferring things from it and that can lead to a solution that doesn't even really solve the problem for them or solves it in a non-ideal way.

The only reason that this change was made by @Araq is because of the assumption that this affects the forum, but it very clearly does not, so we should seriously reevaluate here. Assuming things is how you end up with badly designed software and yes, I am stubborn about avoiding that.

Reply:
It solves the simplified version of @mrhdias's problem I provided, as explained about 6 times now, with a chain of simplifications starting from the Forum, on the issue, on the 1st PR and this PR.

I think you have offered zero actual rebuttals other than "robustness bad because it masks errors", and I have explained why that is a weak rebuttal since people looking for such errors should take special action anyway.  So, zero valid rebuttals so far.  There may even be a valid one, but you haven't mentioned one.

I think you are reading way too much into @Araq's motivations when you say "only reason", but I will let him speak for himself on this from now on.

Complexity-wise this is no different than checking you have 100 GiB free disk space before writing a 100 GiB file.